### PR TITLE
Add `TryFrom` to convert repr to enum

### LINKED ISSUE #146: Derive `TryFrom<T>` for enums marked `#[repr(T)]`
C-like enums with an explicit `#[repr(<int type>)]` attribute can be cast to that integer type, but there's no easy way to convert back:

```rust
#[repr(i16)]
enum MyEnum {
    Foo = 1,
    Bar = 2,
}

let val = MyEnum::Bar as i16;
let my_enum: MyEnum = // convert `val` back to `MyEnum`
```

This derive exists in [`enum-utils`](https://docs.rs/enum-utils/0.1.2/enum_utils/derive.TryFromRepr.html) but it would be nice not to have to reach for multiple crates.

### PR DESCRIPTION
**Pull request by ModProg**
Resolves #146

## Synopsis

Conversion from the repr integer to an enum.

## Solution

A `TryFrom` derive macro implementing `TryFrom<integertype> for Enum`.

## Checklist

- [x] Documentation is updated (if required)
- [x] Tests are added/updated (if required)
    - [x] basic tests
    - [x] compile fail
- [x] [CHANGELOG entry][l:1] is added (if required)

[l:1]: /CHANGELOG.md

### CODE REVIEW

File: impl/doc/try_from.md
Comment (ModProg) [inline impl/doc/try_from.md:35]: @JelteF I'm really not sure how to do this better, the problem is, that 1.65 doesn't allow discriminants on enums with fields.

File: impl/src/try_from.rs
Comment (ModProg) [inline impl/src/try_from.rs:67]: There is one todo here, we should also support:
```rs
#[repr(C, u8)]

#[repr(u8, align(4))]

#[repr(align(4))]
#[repr(u8)]
```
etc.

File: impl/src/try_from.rs
Comment (ModProg) [inline impl/src/try_from.rs:20]: This is probably overly conservative, we should be able to assume that it is an `isize` if nothing was specified:

> Under the [default representation](https://doc.rust-lang.org/reference/type-layout.html#the-default-representation), the discriminant is interpreted as an isize value. However, the compiler is allowed to use a smaller type (or another means of distinguishing variants) in its actual memory layout.

https://doc.rust-lang.org/reference/items/enumerations.html#discriminants

File: src/convert.rs
Comment (ModProg) [inline src/convert.rs:79]: Not sure what good error message would be here.

**Comment by ModProg** [comment]
We could also support some `#[try_from(<type>)]` for types whose definition is a superset to the repr type.

File: impl/src/try_from.rs
Comment (ModProg) [reply impl/src/try_from.rs:20]: Got another confirmation on the `isize` for the expression especially: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/enum.20discriminant.20size/near/388599445

File: impl/doc/try_from.md
Comment (tyranron) [reply impl/doc/try_from.md:36]: @ModProg @JelteF I think it's worth bumping MSRV up to 1.66.

Comment (ModProg) [reply impl/doc/try_from.md:36]: On the one hand it's only documentation, on the other hand, it's only one version.

File: tests/compile_fail/try_from/invalid_repr.rs
Comment (tyranron) [inline tests/compile_fail/try_from/invalid_repr.rs:3]: @ModProg it would also be nice to cover the case with _valid_ `repr`, but not meaningful, like `repr(Rust)` or `repr(C)`.

I think, we should control the exact list of types we accept, inside the macro: `i8`, `u8`, `i16`, `u16`, `i32`, `u32`, `i64`, `u64`, `i128`, `u128`, `isize`, `usize`. And disregard anything else.

File: src/convert.rs
Comment (tyranron) [inline src/convert.rs:71]: @ModProg that's too limiting bound. Usually, most of types don't impl `Display`, and thus we make the error here non-`Display`able for the majority of cases. Better remake the error message the way we lift this trait bound. I think adding a type name as `&'static str` to the error would do the trick, as it does for the `TryIntoError` above.

File: impl/doc/try_from.md
Comment (tyranron) [inline impl/doc/try_from.md:30]: @ModProg this very rare and uncommon situation, so I think library users deserve some reasoning and explanation why this is not supported exactly, because it's quite unclear, imho.

File: impl/src/try_from.rs
Comment (tyranron) [inline impl/src/try_from.rs:46]: > I think, we should control the exact list of types we accept, inside the macro: `i8`, `u8`, `i16`, `u16`, `i32`, `u32`, `i64`, `u64`, `i128`, `u128`, `isize`, `usize`. And disregard anything else.

@ModProg you do so, already. Good 👍

So we need only a test covering that.

File: impl/src/try_from.rs
Comment (tyranron) [inline impl/src/try_from.rs:60]: @ModProg let's mention this in docs too. Quite unclear and subtle moment.

File: impl/src/try_from.rs
Comment (tyranron) [inline impl/src/try_from.rs:73]: @ModProg I think it's not worth here to bother with lifetimes and complicating signatures. Just `Clone`/`Copy` is enough. It won't give any real performance benefit here, so the library users can feel it.

File: impl/doc/try_from.md
Comment (JelteF) [reply impl/doc/try_from.md:36]: I would say let's just bump it to 1.72, by the time we release 1.0 I think that will have been out long enough.

File: impl/src/try_from.rs
Comment (tyranron) [inline impl/src/try_from.rs:81]: @ModProg it's better and more natural to write expansion as `ToTokens` impl. This also prevents the expansion to be fallible. Anything that can error should be checked before performing the expansion.

**Review by tyranron — CHANGES_REQUESTED** [review]
@ModProg good job! Thanks 👍
Overall design is solid and thoughtful. The implementation is also good, despite few bikesheddings described below.

> We could also support some `#[try_from(<type>)]` for types whose definition is a superset to the repr type.

Definitely not as a part of this PR. Let's sketch and discuss the design of it in #146.

File: src/convert.rs
Comment (ModProg) [reply src/convert.rs:71]: Maybe it would make sense to have an explicit `TryFromRepr` error, then? Because all types available as repr do in fact implement `Display`.

File: impl/doc/try_from.md
Comment (ModProg) [reply impl/doc/try_from.md:30]: Maybe:
> Variants with fields are not supported, as the value for their fields would be undefined.

File: impl/src/try_from.rs
Comment (ModProg) [reply impl/src/try_from.rs]: https://github.com/JelteF/derive_more/blob/819cb3d2b98e08c941b4747b3df70d4d15aceca2/impl/doc/try_from.md?plain=1#L4-L5

File: impl/src/try_from.rs
Comment (ModProg) [reply impl/src/try_from.rs:74]: Probably true, I took this and the fallible `expand` from the `From` implementation

**Comment by tyranron** [comment]
@ModProg @JelteF yet another thing that i think of: we will definitely want to support this case in the future:
```rust
#[derive(TryFrom)]
enum Foo {
    #[try_from(u128)] // generates `impl TryFrom<u128> for Foo`
    Bar(u8),
}
```
While at the same time, in other situations we would like to have something like this:
```
#[derive(TryFrom)]
#[try_from(u128)] // generates `impl TryFrom<u128> for Foo`
#[repr(u8)]
enum Foo {
    Bar,
}
```
So, we need a way to differentiate a `repr` conversions from the ones of the inner types clearly.

And this question touches not only `TryFrom`, because we also will want to have something like this for `Into`:
```rust
#[derive(Into)] // generates `impl From<Foo> for u8`
#[repr(u8)]
enum Foo {
    Bar = 35,
}
```
And `TryInto`:
```rust
#[derive(TryInto)]
#[try_into(u8)] // generates `impl TryFrom<Foo> for u8`
#[repr(u128)] 
enum Foo {
    Bar = 35,
}
```

To disambiguate them clearly, I propose to use `repr` sub-attribute:
```rust
#[derive(Into, TryFrom)]
#[into(repr)] // generates `impl From<Foo> for u8`
#[try_from(repr)] // generates `impl TryFrom<u8> for Foo`
#[repr(u8)]
enum Foo1 {
    Bar,
}

#[derive(TryFrom, TryInto)]
#[try_from(repr(u128))] // generates `impl TryFrom<u128> for Foo`
#[try_into(repr(u8))] // generates `impl TryFrom<Foo> for u8`
#[repr(u16)]
enum Foo2 {
    Bar,
}
```

Does it make sense?

**Comment by ModProg** [comment]
@tyranron, looks sensible. I'll implement support for the repr syntax for now, and will error on it not specified.

**Comment by ModProg** [comment]
@tyranron how would you expect `repr(<types>)` to work?

```
impl TryFrom<$type> for $enum {
    fn try_from(value: $type) -> Result<Self> {
        let value: $actual_repr = value.try_into().map_err(|e| TryFromError::message(e.to_string())?;
        match value {
            $actual_implementation
        }
    }
}
```

**Comment by tyranron** [comment]
@ModProg as a regular forwarding implementation, like you've pointed, yes. But as I've [mentioned before](https://github.com/JelteF/derive_more/pull/300#pullrequestreview-1612916424), let's not making it as a part of this PR. Better implement as a separate one after merging this.

File: impl/src/try_from.rs
Comment (tyranron) [reply impl/src/try_from.rs:46]: @ModProg seems to be enough

File: src/convert.rs
Comment (tyranron) [reply src/convert.rs:71]: @ModProg yes, makes sense.

File: impl/doc/try_from.md
Comment (JelteF) [inline impl/doc/try_from.md:19]: I think the example is good, but hard to understand at a glance.
```suggestion
    ImplicitZero,
    ExplicitFive = 5,
    FieldSix(usize),
    EmptySeven{},
```

File: src/convert.rs
Comment (JelteF) [inline src/convert.rs:71]:
```suggestion
// `T` should only be an integer type and therefore be debug
```

File: src/convert.rs
Comment (JelteF) [inline src/convert.rs:74]:
```suggestion
        write!(f, "`{:?}` does not corespond to a unit variant", self.input)
```

**Review by JelteF — DISMISSED** [review]
Looks good to me (apart from the few small comments I left). @tyranron did you have any more feedback.

File: impl/doc/try_from.md
Comment (ModProg) [inline impl/doc/try_from.md:24]:
```suggestion
assert_eq!(Enum::ImplicitZero, Enum::try_from(0).unwrap());
assert_eq!(Enum::ExplicitFive, Enum::try_from(5).unwrap());
assert_eq!(Enum::EmptySeven{}, Enum::try_from(7).unwrap());
```

**Comment by tyranron** [comment]
@JelteF yes, I was waiting for #298 to land. Give me a moment.

**Comment by tyranron** [comment]
@JelteF oh, and please, change the required CI jobs in project settings to reflect MSRV 1.65.0 -> MSRV 1.72.0 migration. Otherwise, CI won't pass. Thanks!

**Review by tyranron — APPROVED** [review]

**Pull request merged**
